### ISSUE TITLE
MySQL 5.7 Unsupported

### ISSUE THREAD

**Issue by TheSerapher** [MPOS/php-mpos#2577]
While working on a local dev setup I noticed that MPOS is not ready for MySQL 5.7. Some defaults have changed that will cause things like payouts to not find any valid shares. The error itself is caused by a changed globals setting in MySQL.

For now, either run an older MySQL Version (you will have to try out which one works) or follow this hack to disable the new option and return to the old behaviour: https://stackoverflow.com/a/35729681

I implemented a fix, but I am not confident that it's going to work on an older SQL version without causing any issues for the users. Hence please use that workaround for now.

**Comment by TheSerapher** [comment]
Upon further investigation there are quite a few places that are affected by the new `GROUP BY` policy set via `sql_mode=only_full_group_by`. I have tried to fix https://github.com/MPOS/php-mpos/blob/master/include/classes/statistics.class.php#L183-L190 as an example but it doesn't work properly. It seems that I may need assistance by someone more familiar with this change and how this can be addresses without breaking MPOS :)

**Comment by TheSerapher** [comment]
Here the example used. Even though this resulted in a cryptic error message when doing payouts it didn't break anything beyond that. I am hesitant to just change the query as suggested because in this example it does return wrong results. In a payment processor this could be catastrophic!

**Enable Workaround, execute SQL**
```
mysql> set global sql_mode='STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION';
Query OK, 0 rows affected (0.00 sec)

mysql> set session sql_mode='STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION';
Query OK, 0 rows affected (0.00 sec)

mysql> SELECT         b.*,         a.username AS finder,         a.is_anonymous AS is_anonymous,         COUNT(b.id) AS solvedblocks,         SUM(b.amount) AS generatedcoins       FROM blocks AS b       LEFT JOIN accounts AS a       ON b.account_id = a.id       WHERE confirmations > 0       GROUP BY finder       ORDER BY solvedblocks DESC LIMIT 10;
+----+--------+------------------------------------------------------------------+---------------+--------+-----------------------+------------+-----------+------------+--------------------+--------+----------+-------------+--------------+--------------+----------------+
| id | height | blockhash                                                        | confirmations | amount | difficulty            | time       | accounted | account_id | worker_name        | shares | share_id | finder      | is_anonymous | solvedblocks | generatedcoins |
+----+--------+------------------------------------------------------------------+---------------+--------+-----------------------+------------+-----------+------------+--------------------+--------+----------+-------------+--------------+--------------+----------------+
|  1 | 118821 | 1f8b06c1198d4be1bb369cd10e2a160b391ff02e35d2854853c0482be02eb82e |           172 |     50 | 0.0002441371325370145 | 1500453302 |         1 |          1 | TheSerapher.worker |      9 |       56 | TheSerapher |            0 |            4 |            200 |
+----+--------+------------------------------------------------------------------+---------------+--------+-----------------------+------------+-----------+------------+--------------------+--------+----------+-------------+--------------+--------------+----------------+

```

**Disable Workaround, execute SQLs**
```
mysql> set global sql_mode='STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION,ONLY_FULL_GROUP_BY';
Query OK, 0 rows affected (0.00 sec)

mysql> set session sql_mode='STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION,ONLY_FULL_GROUP_BY';
Query OK, 0 rows affected (0.00 sec)

mysql> SELECT         b.*,         a.username AS finder,         a.is_anonymous AS is_anonymous,         COUNT(b.id) AS solvedblocks,         SUM(b.amount) AS generatedcoins       FROM blocks AS b       LEFT JOIN accounts AS a       ON b.account_id = a.id       WHERE confirmations > 0       GROUP BY finder       ORDER BY solvedblocks DESC LIMIT 10;
ERROR 1055 (42000): Expression #1 of SELECT list is not in GROUP BY clause and contains nonaggregated column 'mpos_ltc.b.id' which is not functionally dependent on columns in GROUP BY clause; this is incompatible with sql_mode=only_full_group_by
mysql> SELECT         b.*,         a.username AS finder,         a.is_anonymous AS is_anonymous,         COUNT(b.id) AS solvedblocks,         SUM(b.amount) AS generatedcoins       FROM blocks AS b       LEFT JOIN accounts AS a       ON b.account_id = a.id       WHERE confirmations > 0       GROUP BY finder, b.id       ORDER BY solvedblocks DESC LIMIT 10;
+----+--------+------------------------------------------------------------------+---------------+--------+-----------------------+------------+-----------+------------+--------------------+--------+----------+-------------+--------------+--------------+----------------+
```
**Same as above with adjusted `GROUP BY` with wrong result**
```
| id | height | blockhash                                                        | confirmations | amount | difficulty            | time       | accounted | account_id | worker_name        | shares | share_id | finder      | is_anonymous | solvedblocks | generatedcoins |
+----+--------+------------------------------------------------------------------+---------------+--------+-----------------------+------------+-----------+------------+--------------------+--------+----------+-------------+--------------+--------------+----------------+
|  1 | 118821 | 1f8b06c1198d4be1bb369cd10e2a160b391ff02e35d2854853c0482be02eb82e |           172 |     50 | 0.0002441371325370145 | 1500453302 |         1 |          1 | TheSerapher.worker |      9 |       56 | TheSerapher |            0 |            1 |             50 |
|  2 | 118817 | 8ca7cc19d8f6edcce088bd0c50187533bdde1b642d3873b9b5ac6710dcb14bb2 |           176 |     50 | 0.0002441371325370145 | 1500453076 |         1 |          1 | TheSerapher.worker |     40 |       37 | TheSerapher |            0 |            1 |             50 |
|  3 | 118823 | ef5bb49a9902db7e4d9691331df2763bf7aabcc242af94b38844bb5b949ed787 |           170 |     50 | 0.0002441371325370145 | 1500453365 |         1 |          1 | TheSerapher.worker |      1 |       57 | TheSerapher |            0 |            1 |             50 |
|  4 | 118820 | eb338480b8eeb08f61ba7459c5ea183689635f21970463c52b6b5d9c39f4683a |           173 |     50 | 0.0002441371325370145 | 1500453262 |         1 |          1 | TheSerapher.worker |     21 |       50 | TheSerapher |            0 |            1 |             50 |
+----+--------+------------------------------------------------------------------+---------------+--------+-----------------------+------------+-----------+------------+--------------------+--------+----------+-------------+--------------+--------------+----------------+
```

**Comment by TheSerapher** [comment]
I looked at more queries, for a SQL specialist this may be easy to solve, for me not having worked on SQL or MPOS for such a long time it may not be feasible to get it done in a proper amount of time. Please use the workaround, I have added it to the FAQ for the time being: https://github.com/MPOS/php-mpos/wiki/FAQ#is-mysql-57-supported

**Comment by zimbo123** [comment]
Hello,

i have the same Problem.
But i didn't understand the Workaround.
I wrote in the SQL-Consle of the MPOS-Database

set global sql_mode='STRICT_TRANS_TABLES,NO_ZERO_IN_DATE,NO_ZERO_DATE,ERROR_FOR_DIVISION_BY_ZERO,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION';

But the Problem wasn't solved. Where is my mistake?
Who can help me?
Sorry, but my SQL-KnowHow is very low.
Thank you very much.

**Comment by m1dnight** [comment]
Which version is sure to work?

**Comment by zimbo123** [comment]
MySQL 5.6 and 5.5 works fine.

**Comment by TheSerapher** [comment]
I added a MySQL version check to the admin setup page. Since I am unsure of anything > 5.7 it's just checking `5.7` for now anywhere in the version string returned by a SQL query (`SELECT VERSION()`). Only available in `development` or `MPOS > 1.1.0`.

If this is an issue later down the road we may have to adjust the check, but it'll help for a few first time setups at least.
